libcudacxx/include/cuda/std/detail/libcxx/include/support/atomic/atomic_cuda_local.h

```diff
@@ -0,0 +1,191 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+#ifndef __LIBCUDACXX_ATOMIC_CUDA_LOCAL_H
+#define __LIBCUDACXX_ATOMIC_CUDA_LOCAL_H
+
+_LIBCUDACXX_HOST_DEVICE
+inline int __cuda_memcmp(void const * __lhs, void const * __rhs, size_t __count) {
+    NV_DISPATCH_TARGET(
+        NV_IS_DEVICE, (
+            unsigned char const * __lhs_c;
+            unsigned char const * __rhs_c;
+            // NVCC recommended laundering through inline asm to compare padding bytes.
+            asm ("mov.b64 %0, %2;\n mov.b64 %1, %3;" : "=l"(__lhs_c), "=l"(__rhs_c) : "l"(__lhs), "l"(__rhs));
+            while (__count--) {
+                auto const __lhs_v = *__lhs_c++;
+                auto const __rhs_v = *__rhs_c++;
+                if (__lhs_v < __rhs_v) { return -1; }
+                if (__lhs_v > __rhs_v) { return 1; }
+            }
+            return 0;
+        ),
+        NV_IS_HOST, (
+            return memcmp(__lhs, __rhs, __count);
+        )
+    )
+}
+
+// This file works around a bug in CUDA in which the compiler miscompiles
+// atomics to automatic storage (local memory). This bug is not fixed on any
+// CUDA version yet.
+//
+// CUDA compilers < 12.3 also miscompile __isLocal, such that the library cannot
+// detect automatic storage and error. Therefore, in CUDA < 12.3 compilers this
+// uses inline PTX to bypass __isLocal.
+
+_LIBCUDACXX_DEVICE inline bool __cuda_is_local(const void* __ptr) {
+#if defined(_LIBCUDACXX_CUDACC_BELOW_12_3)
+  int __tmp = 0;
+  asm(
+      "{\n\t"
+      "  .reg .pred %p;\n\t"
+      "  isspacep.local %p, %1;\n\t"
+      "  @%p mov.s32 %0, 1;\n\t"
```

Comment:
important: nvcc accepts variable names starting with `%`, but clang does not:
```suggestion
      "  .reg .pred p;\n\t"
      "  isspacep.local p, %1;\n\t"
      "  @p mov.s32 %0, 1;\n\t"
```